scylla/src/transport/load_balancing/default.rs

```diff
@@ -1,25 +1,334 @@
 use super::{FallbackPlan, LoadBalancingPolicy, NodeRef, QueryInfo};
-use crate::transport::cluster::ClusterData;
+use crate::{
+    routing::Token,
+    transport::{cluster::ClusterData, node::Node, topology::Strategy},
+};
+use itertools::{Either, Itertools};
+use rand::{
+    prelude::{IteratorRandom, SliceRandom},
+    thread_rng, Rng,
+};
+use scylla_cql::{frame::types::SerialConsistency, Consistency};
+use std::sync::Arc;
 
+/// This is the default load balancing policy.
+/// It can be configured to be datacenter-aware and token-aware.
+/// Datacenter failover for queries with non local consistency mode is also supported.
 #[derive(Debug)]
-pub struct DefaultPolicy {}
+pub struct DefaultPolicy {
+    preferred_datacenter: Option<String>,
+    is_token_aware: bool,
+    permit_dc_failover: bool,
+}
 
-impl DefaultPolicy {
-    pub fn new() -> Self {
-        Self {}
+impl LoadBalancingPolicy for DefaultPolicy {
+    fn pick<'a>(&'a self, query: &'a QueryInfo, cluster: &'a ClusterData) -> NodeRef<'a> {
```

Comment:
I've got a feeling that `pick` and `fallback` methods follow a similar pattern (control flow-wise) and we could try to reduce the complexity by deduplicating them. Basically, you could think of `pick` as a function that returns the first element of `fallback` without having to allocate memory for the `FallbackPlan`. Of course, if we implemented `pick` such that it calls` fallback` then the performance benefit would be gone - however, what if we removed the need to allocate the plan with a callback?

You could implement a function `do_with_plan` (didn't think too much about the name):

```rust
// FnOnce won't be sufficient to express callback's type, because you don't have
// access to the iterator's type - you will most likely have to use a trait
trait PlanExtractor<'a> {
    type Output;
    fn extract<I>(plan: I) -> Self::Output
    where
        I: Iterator<Output = NodeRef<'a> + Send + Sync + 'a>;
}

fn do_with_plan<'a, E>(
    &'a self,
    query: &'a QueryInfo,
    cluster: &'a ClusterData,
    extractor: E,
) -> E::Output
where
    E: PlanExtractor<'a>,
{
    /* snip */

    let plan = maybe_replicas
        .chain(robined_local_nodes)
        .chain(maybe_remote_nodes)
        .unique();

    extractor.extract(plan)
}
```

I didn't really scrutinize the `pick` and `fallback` methods so I'm not sure what the performance impact of this approach will be. If computing the plan requires significantly more computations than the current implementation of `pick` (apart from avoiding the allocation), then I guess we can stay with the current approach.

It would be great if we could compare the performance of both approaches. Perhaps a benchmark would be in order? After all, one of the reasons for the refactor was to improve the performance.